# Домашняя работа №2

**Срок сдачи:** 21 октября 2017, 23:59 <br\>
Штраф за опоздание: -1 балл за каждый день

Файл с дз надо загрузить по [ссылке](https://www.dropbox.com/request/Sc8LELV7lp5Vl3Ad7wJG)<br/>
Постарайтесь назвать файл по следующему формату:** [ИАД-*{Номер}*] *{Фамилия}* *{Имя}* ДЗ*{Номер}* **<br/>

Сопровождайте ваш код изображеними, комментариями и выводами. <br/>
Иммейте ввиду, что на некоторые задачи нет единственного верного и полного ответа. Чем больше информации вы сможете извлечь, аргументированных выводов сформулировать, тем лучше.

Используйте данный Ipython Notebook при оформлении домашнего задания.

# Table of Contents
 <p><div class="lev1 toc-item"><a href="#Домашняя-работа-2" data-toc-modified-id="Домашняя-работа-2-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Домашняя работа 2</a></div><div class="lev2 toc-item"><a href="#Работа-с-данными-(4-балла)" data-toc-modified-id="Работа-с-данными-(4-балла)-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Работа с данными (4 балла)</a></div><div class="lev2 toc-item"><a href="#Классификация-(6-баллов)" data-toc-modified-id="Классификация-(6-баллов)-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Классификация (6 баллов)</a></div><div class="lev3 toc-item"><a href="#Предписания" data-toc-modified-id="Предписания-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Предписания</a></div><div class="lev3 toc-item"><a href="#Задание" data-toc-modified-id="Задание-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Задание</a></div>

Прочтите статью ["USING	DATA	MINING	TO	PREDICT	SECONDARY SCHOOL	STUDENT	ALCOHOL	CONSUMPTION"](https://www.dropbox.com/s/ww4h9ivnkbyy9xw/STUDENT%20ALCOHOL%20CONSUMPTION%20%281%29.pdf?dl=0).<br\>
Загрузите [датасет](https://cloud.mail.ru/public/KiKi/Cow8y2yn7), используемый в статье и выполните следующие задания:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df = pd.read_csv('./student/student-por.csv', sep=';')

In [3]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other ...   
3     GP   F   15       U     GT3       T     4     2   health  services ...   
4     GP   F   16       U     GT3       T     3     3    other     other ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

## Работа с данными (4 балла)
1. Приведите описание признаков датасета на русском языке с указанием типа данных (2 балла)
2. Описание предобработки данных (2 балла)
    - Проверьте наличие пропусков. В случае наличия пропусков заполните их медианными значениями (1 балл)
    - Преобразуйте все номинальные признаки в несколько признаков с бинарным значением (1 балл)

### 1. Описание признаков датасета

Честно говоря, не очень хорошо понимаю смысл этого задания, особенно после слов Андрея на семинаре о том, что перевод не нужен - "Вы же не переводчики". Описание данных приведено в статье, причем в достаточно удобном виде. Ну да ладно, раз уж сказано описать признаки датасета "на русском языке", так и поступлю.

Сначала посмотрим, какие вообще признаки есть. Это очень легко, так как они являются названиями колонок в датасете.

In [4]:
df.columns, len(df.columns)

(Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
        'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
        'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
        'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
        'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
       dtype='object'), 33)

Для удобства помещу признаковое описание из статьи здесь

![title](prizn.png)

Я хочу создать новый датафрейм descr_features, включающий название признака, его тип и описание.

Это не просто способ удобно представить информацию, но и возможность быстро выводить информацию об интересущих признаках далее в работе. Не знаю, понадобится это или нет, но лишним не будет

In [5]:
features = df.columns

descr_features = pd.DataFrame(columns=['feature', 'type', 'description'])

descr_features.feature = features
descr_features.type = 'числовой'

Добавляем бинарные и номинальные значения

In [6]:
binary_features = ['school','sex','address','famsize','Pstatus','schoolsup',
                   'famsup','paid','activities', 'nursery','higher',
                   'internet','romantic']

nominal_features = ['Mjob', 'Fjob', 'reason', 'guardian']

In [7]:
i = 0
for feature in descr_features.feature:
    if feature in binary_features:
        descr_features.type[i] = 'бинарный'
    if feature in nominal_features:
        descr_features.type[i] = 'номинальный'
    i += 1

Осталось описание... Ну как-то странно и долго, потому что писать придется вручную и на это уйдет много времени. Ну да ок

Я заметила, что не все переменные в нашем датасете и описании, представленном в статье, совпадают. Например, в статье учитывается только образование матери, а про отца ничего не спрашивают. Смахивает на дискриминацию, но в наших данных информация об отцах тоже есть. Поэтому в следующей ячейке будут описаны те переменные, которые есть в статье + те, которые являются идентичными, но в статье не указаны (если есть информация о матерях, а в нашем датасете есть переменная, учитывающая эту же сторону жизни отцов, она тоже будет в следующей ячейке).

In [8]:
descr_features.description[0] = 'Школа: Gabriel Pereira или Mousinho da Silveira'
descr_features.description[1] = 'Пол: женский или мужской'
descr_features.description[2] = 'Возраст: от 15 до 22'
descr_features.description[3] = 'Тип домашнего адреса: живет в городе или сельской местности'

descr_features.description[5] = 'Статус сожительства родителей: живут вместе или раздельно'
descr_features.description[6] = 'Образование матери: 0 - нет, 1 - начальное, 2 - с 5 по 9 класс,\
                                    3 - старшая школа, 4 - высшее образование'
descr_features.description[7] = 'Образование отца: 0 - нет, 1 - начальное, 2 - с 5 по 9 класс,\
                                    3 - старшая школа, 4 - высшее образование'
descr_features.description[8] = 'Работа матери'
descr_features.description[9] = 'Работа отца'


descr_features.description[12] = """Время, которое тратится на поездку из дома в школу: 1 - меньше 15 минут,\ 
                                2 - 15-30 минут, 3 - полчаса-час, 4 - 1 час"""
descr_features.description[13] = 'Сколько часов в неделю учится: 1 - меньше 2 часов, 2 - от 2 до 5 часов,\
                                3 - от 5 до 10, 4 - больше 10'
descr_features.description[14] = 'Количество "провалов" в школе в прошлом: от 1 до 4'
descr_features.description[15] = 'Поддержка школы в вопросах дополнительного образования: да/нет'
descr_features.description[16] = 'Поддержка семьи в вопросах образования: да/нет'
descr_features.description[17] = 'В описании есть похожая переменная, скорее всего: пришлось ли платить\
                                за дополнительное (?) образование - да/нет'
descr_features.description[18] = 'Внеклассные занятия: да/нет'
descr_features.description[19] = 'Был ли в детском саду: да/нет'
descr_features.description[20] = 'Хочет получить высшее образование: да/нет'
descr_features.description[21] = 'Наличие доступа к Интернету дома: да/нет'
descr_features.description[22] = 'Состоит в романтических отношениях: да/нет'

descr_features.description[24] = 'Свободное время после школы: 1 - очень мало, 5 - очень много'
descr_features.description[25] = 'Прогулки с друзьями: 1 - очень редко, 5 - очень часто'

descr_features.description[28] = 'Текущее состояние здоровья: 1 - очень плохое, 5 - очень хорошее'
descr_features.description[29] = 'Количество прогулов в школе: от 0 до 93'
descr_features.description[30] = 'от 0 до 20'
descr_features.description[31] = 'от 0 до 20' #не оч понятно, то ли оценка, то ли количество выпусков, то ли еще что
descr_features.description[32] = 'от 0 до 20'


Теперь опишем переменные, о которых ничего не сказано в статье

In [9]:
descr_features[4:5]

feature      type description
4  famsize  бинарный         NaN

In [10]:
print(df.famsize.unique())

['GT3' 'LE3']


Похоже, это означает "больше трех" и "меньше или равно трем". Очевидно, речь идет о количестве членов семьи. На самом деле я не уверена, что "LE" - "меньше или равно", так как это может быть и "LEss" и "Less Equal". Более вероятным кажется второй вариант.

Признак бинарный, поэтому нужно добавить его в список binary_features. Можно сделать это через append, но тогда придется копировать ячейку, меняющую тип переменной в датасете на нужный. Поэтому просто скопипастю названия нужных переменных в списки.


Посмотрим на следующие признаки:

In [11]:
descr_features[10:12]

feature         type description
10    reason  номинальный         NaN
11  guardian  номинальный         NaN

In [12]:
print('reason:', df.reason.unique())
print('guardian:', df.guardian.unique())

reason: ['course' 'other' 'home' 'reputation']
guardian: ['mother' 'father' 'other']


Переменные номинальные, поэтому нужно добавить их в список nominal_features

Reason - причина. Вероятно, чтобы понять, про причину чего спрашивают, нужно посмотреть на предыдущую переменную

In [13]:
descr_features[9:10]

feature         type  description
9    Fjob  номинальный  Работа отца

In [14]:
df.Fjob.unique()

array(['teacher', 'other', 'services', 'health', 'at_home'], dtype=object)

Вероятно, это притянуто за уши, но все же мне кажется, что переменная reason - причина, по которой человек работает там, где он работает. Мне не очень понятно, к чему относится, например, 'course', но в целом логично, что, например, человек работает учителем, потому что эта профессия с хорошей репутацией. Но я не уверена :С

Проблема в том, что "работа" есть как у матери, так и у отца, а переменная причины только одна. В общем странная какая-то.

Может быть стоит вообще удалить эту переменную...

In [15]:
Fteacher = df[df.Fjob == 'teacher']
Fteacher[['Mjob', 'Fjob', 'reason']].head()

Mjob     Fjob      reason
0    at_home  teacher      course
7      other  teacher        home
29   teacher  teacher        home
37     other  teacher  reputation
42  services  teacher      course

Все же не очень понятно, что значит эта причина

Guardian переводится как опекун, попечитель. Я искренне не понимаю, почему эта переменная вообще существует. Разве не может быть такого, что оба родителя обеспечивают ребенка в равной степени? Есть вероятность, что такой вариант учтен в other, но это как-то сомнительно. Почему нет инфыыыыы ааааааарррр

In [16]:
df.guardian.describe()

count        649
unique         3
top       mother
freq         455
Name: guardian, dtype: object

Посмотрим на последнюю неописанную переменную

In [17]:
descr_features[23:24]

feature      type description
23  famrel  числовой         NaN

In [18]:
print(df.famrel.unique())

[4 5 3 1 2]


Во-первых, совсем не понятно, что это за цифры. Эта переменная может быть номинальной, а может быть и числовой, как, например, в случае со временем, которое тратится на поездку из дома в школу. Мне не приходит в голову, что означает "rel". Вообще наверное это связано с семейными отношениями (family relations), но я по-прежнему не знаю, что означают цифры. Можно предположить, что 1 - очень плохие, 5 - очень хорошие, но я не могу знать это наверняка.

Рассмотрим еще 2 переменные. В таблице с описанием в статье их нет, но информация есть. Вместе они образуют целевую переменную alc - потребление алкоголя в течение недели, 1 - очень мало, 5 - очень много.

In [19]:
descr_features[26:28]

feature      type description
26    Dalc  числовой         NaN
27    Walc  числовой         NaN

In [20]:
df.Dalc.unique()

array([1, 2, 5, 3, 4])

In [21]:
df.Walc.unique()

array([1, 3, 2, 4, 5])

Попробую написать описание, но я бы не доверяла ему на 100%

In [22]:
descr_features.description[4] = 'Размер семьи: больше 3 человек, меньше 3 человек или 3 человека'
descr_features.description[10] = "Какая-то непонятная причина, варианты ответов: 'course','reputation',\
                                'home', 'other'"
descr_features.description[11] = 'Опекун, попечитель: мать/отец/другое'
descr_features.description[23] = 'Семейные отношения (?): 1 - очень плохие, 5 - очень хорошие'
descr_features.description[26] = 'Сколько алкоголя пьет в рабочие дни'
descr_features.description[27] = 'Сколько алкоголя пьет в выходные'

In [23]:
descr_features

feature         type                                        description
0       school     бинарный    Школа: Gabriel Pereira или Mousinho da Silveira
1          sex     бинарный                           Пол: женский или мужской
2          age     числовой                               Возраст: от 15 до 22
3      address     бинарный  Тип домашнего адреса: живет в городе или сельс...
4      famsize     бинарный  Размер семьи: больше 3 человек, меньше 3 челов...
5      Pstatus     бинарный  Статус сожительства родителей: живут вместе ил...
6         Medu     числовой  Образование матери: 0 - нет, 1 - начальное, 2 ...
7         Fedu     числовой  Образование отца: 0 - нет, 1 - начальное, 2 - ...
8         Mjob  номинальный                                      Работа матери
9         Fjob  номинальный                                        Работа отца
10      reason  номинальный  Какая-то непонятная причина, варианты ответов:...
11    guardian  номинальный               Опекун, попечитель: мать/отец/другое
12  traveltime     числовой  Время, которое тратится на поездку из дома в ш...
13   studytime     числовой  Сколько часов в неделю учится: 1 - меньше 2 ча...
14    failures     числовой  Количество "провалов" в школе в прошлом: от 1 ...
15   schoolsup     бинарный  Поддержка школы в вопросах дополнительного обр...
16      famsup     бинарный     Поддержка семьи в вопросах образования: да/нет
17        paid     бинарный  В описании есть похожая переменная, скорее все...
18  activities     бинарный                        Внеклассные занятия: да/нет
19     nursery     бинарный                      Был ли в детском саду: да/нет
20      higher     бинарный          Хочет получить высшее образование: да/нет
21    internet     бинарный           Наличие доступа к Интернету дома: да/нет
22    romantic     бинарный         Состоит в романтических отношениях: да/нет
23      famrel     числовой  Семейные отношения (?): 1 - очень плохие, 5 - ...
24    freetime     числовой  Свободное время после школы: 1 - очень мало, 5...
25       goout     числовой  Прогулки с друзьями: 1 - очень редко, 5 - очен...
26        Dalc     числовой                Сколько алкоголя пьет в рабочие дни
27        Walc     числовой                   Сколько алкоголя пьет в выходные
28      health     числовой  Текущее состояние здоровья: 1 - очень плохое, ...
29    absences     числовой            Количество прогулов в школе: от 0 до 93
30          G1     числовой                                         от 0 до 20
31          G2     числовой                                         от 0 до 20
32          G3     числовой                                         от 0 до 20

### 2. Описание предоработки данных

#### 2.1. Проверка на наличие пропусков

Вообще в самой статье написано, что пропусков нет С:

Но, конечно, можно проверить

In [24]:
df[df.values == None]

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

Может я не так ищу? Попробуем обнаружить значение, которое точно есть

In [25]:
df[df.values == 'at_home'][1:2]

school sex  age address famsize Pstatus  Medu  Fedu     Mjob   Fjob ...  \
1     GP   F   17       U     GT3       T     1     1  at_home  other ...   

  famrel freetime  goout  Dalc  Walc health absences G1  G2  G3  
1      5        3      3     1     1      3        2  9  11  11  

[1 rows x 33 columns]

In [26]:
df[df.values == 'NaN']

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

In [27]:
df[df.values == 'nan']

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

Можно еще вот так

In [28]:
for col in df.columns:
    print(col, '-', df[col].unique())
    print('-------------')

school - ['GP' 'MS']
-------------
sex - ['F' 'M']
-------------
age - [18 17 15 16 19 22 20 21]
-------------
address - ['U' 'R']
-------------
famsize - ['GT3' 'LE3']
-------------
Pstatus - ['A' 'T']
-------------
Medu - [4 1 3 2 0]
-------------
Fedu - [4 1 2 3 0]
-------------
Mjob - ['at_home' 'health' 'other' 'services' 'teacher']
-------------
Fjob - ['teacher' 'other' 'services' 'health' 'at_home']
-------------
reason - ['course' 'other' 'home' 'reputation']
-------------
guardian - ['mother' 'father' 'other']
-------------
traveltime - [2 1 3 4]
-------------
studytime - [2 3 1 4]
-------------
failures - [0 3 1 2]
-------------
schoolsup - ['yes' 'no']
-------------
famsup - ['no' 'yes']
-------------
paid - ['no' 'yes']
-------------
activities - ['no' 'yes']
-------------
nursery - ['yes' 'no']
-------------
higher - ['yes' 'no']
-------------
internet - ['no' 'yes']
-------------
romantic - ['no' 'yes']
-------------
famrel - [4 5 3 1 2]
-------------
freetime - [3 2 4 1

#### 2.2. Преобразование номинальных признаков в несколько признаков с бинарным значением

Какие признаки у нас номинальные? Давайте найдем их в датасете

In [29]:
descr_features[descr_features.type == 'номинальный']

feature         type                                        description
8       Mjob  номинальный                                      Работа матери
9       Fjob  номинальный                                        Работа отца
10    reason  номинальный  Какая-то непонятная причина, варианты ответов:...
11  guardian  номинальный               Опекун, попечитель: мать/отец/другое

Какие значения они могут принимать?

In [30]:
df.Mjob.unique()

array(['at_home', 'health', 'other', 'services', 'teacher'], dtype=object)

In [31]:
df.Fjob.unique()

array(['teacher', 'other', 'services', 'health', 'at_home'], dtype=object)

In [32]:
df.reason.unique()

array(['course', 'other', 'home', 'reputation'], dtype=object)

In [33]:
df.guardian.unique()

array(['mother', 'father', 'other'], dtype=object)

Так как значений у нас больше 2 и группы нельзя объединить так, чтобы получился бинарный признак, придется использовать One-hot кодирование. 

In [34]:
from pandas import get_dummies

In [35]:
df1 = pd.get_dummies(df, columns=['Mjob', 'Fjob', 'reason','guardian'])

In [36]:
df1.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid',
       'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2',
       'G3', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services',
       'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other',
       'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home',
       'reason_other', 'reason_reputation', 'guardian_father',
       'guardian_mother', 'guardian_other'],
      dtype='object')

Но на самом деле не только эти 

In [37]:
df1[1:2]

school sex  age address famsize Pstatus  Medu  Fedu  traveltime  studytime  \
1     GP   F   17       U     GT3       T     1     1           1          2   

        ...        Fjob_other Fjob_services Fjob_teacher reason_course  \
1       ...                 1             0            0             1   

  reason_home reason_other reason_reputation guardian_father guardian_mother  \
1           0            0                 0               1               0   

   guardian_other  
1               0  

[1 rows x 46 columns]

В датасете должны быть только числовые признаки, поэтому бинарные нужно перекодировать в 0 и 1. Можно 

In [38]:
from sklearn.preprocessing import LabelEncoder

In [39]:
le = LabelEncoder()
for feature in binary_features:
    df1[feature] = le.fit_transform(df1[feature])

In [40]:
df1.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid',
       'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2',
       'G3', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services',
       'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other',
       'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home',
       'reason_other', 'reason_reputation', 'guardian_father',
       'guardian_mother', 'guardian_other'],
      dtype='object')

### Могла упустить какие-то номинальные признаки, стоит к этому вернуться позже

----------------------------------------------------------------------------------------------------------------------------

## Классификация (6 баллов)

### Предписания
- Используйте accuracy как основную меру качества
- Классы в задаче несбалансированные. Для корректной кросс-валидации используйте стратифицированный способ разбиения на фолды [Stratified K-fold](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold)

### Задание
1. Задайте RANDOM_SEED и случайным образом разделите выборку на обучающую и контрольную в пропорции 80/20. Этот же RANDOM_SEED используйте при кросс-валидации (0.5 балла)
2. Random Forest (2 балла)
    * В статье описано использование Random Forest для предсказания важности фактором вляющих на потребление алкоголя.<br\>
    * Повторите эксперимент с использованием [RandomForest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier) и найдите наилучшие параметры с помощью кросс-валидации. (1 балл)<br\>
    * Изобразите аналогичную таблицу важности признаков для наилучшей настройки метода (значения могут отличаться). (0.5 балла)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода на контрольной выборке. (0.5 балла)<br\>

3. GradientBoosting (2 балла)
    * Обучите [GradientBoostingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier)  и найдите для него наилучшие параметры с помощью кросс-валидации. (0.5 балла)<br\>
    * Изобразите таблицу важности признаков для наилучшей настройки метода. Отличается ли она от таблицы метода RandomForest. Почему? (1 балл)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода. (0.5 балла)<br\>

4. AdaBoost (1 балл)
    * Обучите [AdaBoostClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html) и найдите для него наилучшие параметры с помощью кросс-валидации. (0.5 балла)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода. (0.5 балла)<br\>

5. Какой из классификаторов оказался лучше? (0.5 балла)

### 1. Задаем random_state и разбиваем выборку

In [41]:
RANDOM_SEED = 42

In [42]:
from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier

$$alc = \frac{Walc*2 + Dalc*5}{7}$$

In [43]:
alc = (df.Walc*2 + df.Dalc*5)/7
alc[alc<3] = 0
alc[alc>=3] = 1
y = alc.astype(int)

In [44]:
df1.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid',
       'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2',
       'G3', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services',
       'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other',
       'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home',
       'reason_other', 'reason_reputation', 'guardian_father',
       'guardian_mother', 'guardian_other'],
      dtype='object')

In [45]:
del df1['Walc']

In [46]:
del df1['Dalc']

In [47]:
X = df1

In [48]:
len(X), len(y), X.shape, y.shape

(649, 649, (649, 44), (649,))

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=RANDOM_SEED)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((519, 44), (130, 44), (519,), (130,))

### 2. Обучаем случайный лес

2. Random Forest (2 балла)
    * В статье описано использование Random Forest для предсказания важности фактором вляющих на потребление алкоголя.<br\>
    * Повторите эксперимент с использованием [RandomForest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier) и найдите наилучшие параметры с помощью кросс-валидации. (1 балл)<br\>
    * Изобразите аналогичную таблицу важности признаков для наилучшей настройки метода (значения могут отличаться). (0.5 балла)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода на контрольной выборке. (0.5 балла)<br\>

Там в статье признак absenses делали бинарным, раз нужно "повторить эксперимент", то я тоже так поступлю

In [50]:
df1['absences'] = (df1['absences'] >= 10).astype(int)

Теперь попробуем обучить случайный лес. Сначала проверю, как он работает на данных, которые в статье названы лучшими, то есть при 25 решающих деревьях.

In [51]:
forest = RandomForestClassifier(n_estimators=25)

In [52]:
y_pred = forest.fit(X_train, y_train).predict(X_test)

In [53]:
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.94615384615384612

Довольно высокое значение accuracy, но в задании сказано найти параметры с помощью кросс-валидации. Следовательно, нам нужно использовать GridSearchCV)

In [75]:
from sklearn.model_selection import GridSearchCV, cross_val_score

In [55]:
from sklearn.metrics import make_scorer

In [96]:
forest = RandomForestClassifier(random_state=RANDOM_SEED)

forest_params = {'n_estimators': range(5, 50, 5), 'max_depth': np.logspace(0, 2, 7, dtype=int)}

forest_grid = GridSearchCV(forest, forest_params, scoring=make_scorer(accuracy_score), cv=25)

forest = forest_grid.fit(X_train, y_train)

forest_grid.best_params_, forest_grid.best_score_

({'max_depth': 4, 'n_estimators': 5}, 0.87475915221579958)

In [97]:
forest_best_est = forest.best_estimator_
forest_best_est.get_params()

{'bootstrap': True,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': 4,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_split': 1e-07,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 5,
 'n_jobs': 1,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

У меня получились вот такие параметры. Это отличается от того, что сказано в статье, но мы используем разные данные, да еще и random_state устанавливаем, поэтому наверное все хорошо.

Посчитаем accuracy

In [58]:
accuracy_grid = accuracy_score(y_test, forest_grid.predict(X_test))
accuracy_grid

0.94615384615384612

Accuracy увеличилось по сравнению с первым разом, когда единственным параметром не по умолчанию было количество деревьев и равнялось оно 25. Сейчас задача решается немного лучше - почти на 0.01. Блин... Теперь я перезапустила ноутбук и они стали одинаковыми :С Не могу понять, что пошло не так) 

In [59]:
accuracy, accuracy_grid, accuracy_grid-accuracy

(0.94615384615384612, 0.94615384615384612, 0.0)

In [115]:
fimp = forest_best_est.feature_importances_
df_fimp_1 = pd.DataFrame(fimp, columns=['важность признака'], index=df1.columns)
df_fimp_1_sorted = df_fimp_1.sort_values('важность признака', ascending=False).head(14)
df_fimp_1_sorted

важность признака
sex                  0.073509
G2                   0.070294
higher               0.067714
traveltime           0.058583
G1                   0.050296
age                  0.048236
failures             0.042929
absences             0.041274
famrel               0.041187
nursery              0.038468
Fedu                 0.038278
G3                   0.037596
address              0.032510
goout                0.031685

Но вообще-то может быть у нас есть какие-то признаки, которые стоит выкинуть, потому что они только мешают и никак не влияют на целевую переменную? Для этого нам понадобятся корреляции каждого признака с целевой переменной

In [60]:
corr_features = []
for feature in df1.columns:
    crr = df1[feature].corr(y)
    print(feature, round(crr, 2))
    if crr > 0.1 or crr < -0.1:
        corr_features.append({feature: round(crr, 2)})

school 0.06
sex 0.25
age 0.14
address -0.07
famsize 0.06
Pstatus 0.0
Medu -0.02
Fedu 0.0
traveltime 0.11
studytime -0.13
failures 0.12
schoolsup -0.04
famsup -0.05
paid 0.03
activities 0.03
nursery -0.05
higher -0.09
internet 0.02
romantic 0.02
famrel -0.04
freetime 0.08
goout 0.2
health 0.05
absences 0.14
G1 -0.18
G2 -0.17
G3 -0.17
Mjob_at_home 0.0
Mjob_health -0.05
Mjob_other -0.01
Mjob_services 0.02
Mjob_teacher 0.03
Fjob_at_home -0.04
Fjob_health 0.01
Fjob_other -0.05
Fjob_services 0.1
Fjob_teacher -0.05
reason_course -0.07
reason_home 0.04
reason_other 0.12
reason_reputation -0.05
guardian_father 0.0
guardian_mother -0.06
guardian_other 0.1


Вообще корреляции очень маленькие, я понятия не имею, откуда они взяли, что корреляция < 0,35 только у 4 признаков, которые решили выкинуть за ненадобностью. Я обучила случайный лес на всех переменных, которые были, и accuracy кажется мне достаточно хорошим.

![title](corr.png)

Поэтому я, в свою очередь, выведу признаки, которые хоть как-то коррелируют. Возможно, проблема в том, что они использовали данные из двух датасетов, но все же это сомнительно. Я работаю с более объемным датасетом, значит по логике, если признаки коррелируют плохо здесь, то при объединении они скорее всего тоже будут коррелировать не очень хорошо.

In [61]:
corr_features

[{'sex': 0.25},
 {'age': 0.14000000000000001},
 {'traveltime': 0.11},
 {'studytime': -0.13},
 {'failures': 0.12},
 {'goout': 0.20000000000000001},
 {'absences': 0.14000000000000001},
 {'G1': -0.17999999999999999},
 {'G2': -0.17000000000000001},
 {'G3': -0.17000000000000001},
 {'reason_other': 0.12},
 {'guardian_other': 0.10000000000000001}]

Признаки, которые авторы статьи решили выкинуть, являются одними из наиболее коррелирующих (относительно всех признаков, конечно, так-то 0,1 - слабая корреляция

### На этом моменте я внезапно узнаю, что дедлайн вчера, а не сегодня. Тадададааммм

Не, ну так-то можно попробовать обучить без этих признаков. Давайте попробуем

In [62]:
df2 = df1.copy()

In [63]:
for col in ['Fedu', 'G1', 'G2', 'G3']:
    del df2[col]

In [68]:
len(df1.columns), len(df2.columns)

(44, 40)

In [69]:
X_new_train, X_new_test, y_new_train, y_new_test = train_test_split(df2, y, train_size=0.8,
                                                                    random_state=RANDOM_SEED)

In [71]:
forest = RandomForestClassifier(random_state=RANDOM_SEED)

forest_params = {'n_estimators': range(5, 50, 5), 'max_depth': np.logspace(0, 2, 7, dtype=int)}

forest_grid = GridSearchCV(forest, forest_params, scoring=make_scorer(accuracy_score), cv=25)

forest = forest_grid.fit(X_new_train, y_new_train)

forest_grid.best_params_, forest_grid.best_score_

({'max_depth': 10, 'n_estimators': 25}, 0.87475915221579958)

Хм, действительно, если сделать как в статье, получится такое же количество решающих деревьев...

In [72]:
forest_best_est = forest.best_estimator_
forest_best_est.get_params()

{'bootstrap': True,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': 10,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_split': 1e-07,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 25,
 'n_jobs': 1,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

In [73]:
accuracy_grid_new = accuracy_score(y_new_test, forest_grid.predict(X_new_test))
accuracy_grid_new

0.93846153846153846

Accuracy ожидаемо стало хуже, но все же ненамного

In [74]:
accuracy, accuracy_grid, accuracy_grid_new

(0.94615384615384612, 0.94615384615384612, 0.93846153846153846)

Теперь нужно сделать табличку важности признаков, как было в статье

In [93]:
df_fimp_2 = pd.DataFrame(forest_best_est.feature_importances_, columns=['важность признака'], index=df2.columns)
df_fimp_2_sorted = df_fimp_2.sort_values('важность признака', ascending=False).head(14)
df_fimp_2_sorted

важность признака
goout                   0.065964
sex                     0.065883
age                     0.064268
famrel                  0.061729
freetime                0.050738
health                  0.044391
Medu                    0.039316
studytime               0.038641
traveltime              0.038542
failures                0.035873
activities              0.029580
reason_other            0.029351
Fjob_services           0.025777
famsup                  0.023360

In [100]:
df_fimp_1.sort_values('важность признака', ascending=False).head(14)

важность признака
sex                  0.073509
G2                   0.070294
higher               0.067714
traveltime           0.058583
G1                   0.050296
age                  0.048236
failures             0.042929
absences             0.041274
famrel               0.041187
nursery              0.038468
Fedu                 0.038278
G3                   0.037596
address              0.032510
goout                0.031685

Все же мне кажется, что не нужно убирать признаки, от которых избавились в статье. У них самих на первом месте стоит пол, у нас такой результат получается в первом случае

Осталось указать значения accuracy для 5 наилучших настроек метода. 

In [201]:
results_1 = pd.DataFrame(forest.cv_results_)
results_1.sort_values(['mean_test_score', 'rank_test_score'], ascending=False).head()

mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
18       0.011108         0.001239         0.874759          0.880943   
43       0.087138         0.005841         0.872832          0.999599   
52       0.097842         0.006577         0.872832          0.999599   
61       0.083107         0.005521         0.872832          0.999599   
0        0.011711         0.001444         0.870906          0.870987   

   param_max_depth param_n_estimators                                  params  \
18               4                  5     {'max_depth': 4, 'n_estimators': 5}   
43              21                 40   {'max_depth': 21, 'n_estimators': 40}   
52              46                 40   {'max_depth': 46, 'n_estimators': 40}   
61             100                 40  {'max_depth': 100, 'n_estimators': 40}   
0                1                  5     {'max_depth': 1, 'n_estimators': 5}   

    rank_test_score  split0_test_score  split0_train_score       ...         \
18                1           0.863636            0.885312       ...          
43                2           0.863636            1.000000       ...          
52                2           0.863636            1.000000       ...          
61                2           0.863636            1.000000       ...          
0                 5           0.863636            0.871227       ...          

    split7_test_score  split7_train_score  split8_test_score  \
18           0.857143            0.883534           0.857143   
43           0.857143            1.000000           0.857143   
52           0.857143            1.000000           0.857143   
61           0.857143            1.000000           0.857143   
0            0.857143            0.871486           0.857143   

    split8_train_score  split9_test_score  split9_train_score  std_fit_time  \
18            0.881526           0.857143            0.883534      0.000826   
43            0.997992           0.857143            1.000000      0.001894   
52            0.997992           0.857143            1.000000      0.017238   
61            0.997992           0.857143            1.000000      0.002648   
0             0.871486           0.857143            0.871486      0.001745   

    std_score_time  std_test_score  std_train_score  
18        0.000023        0.028217         0.003579  
43        0.000756        0.024473         0.000803  
52        0.001487        0.024473         0.000803  
61        0.000927        0.024473         0.000803  
0         0.000341        0.019504         0.000945  

[5 rows x 62 columns]

Я так понимаю, это значения на кросс-валидации, то есть не на отложенной тестовой выборке.

In [234]:
lines = [18, 43, 52, 61, 0]
def best_five_params(lines, df):
    last_params = []
    for i in lines:
        param = df[i:i+1]['params']
        param = param.values
        last_params.append(param)
    return(last_params)
best_five_params(lines=lines, df=results_1)

[array([{'max_depth': 4, 'n_estimators': 5}], dtype=object),
 array([{'max_depth': 21, 'n_estimators': 40}], dtype=object),
 array([{'max_depth': 46, 'n_estimators': 40}], dtype=object),
 array([{'max_depth': 100, 'n_estimators': 40}], dtype=object),
 array([{'max_depth': 1, 'n_estimators': 5}], dtype=object)]

Наверное как-то можно обучить случайный лес с каждым из параметров с помощью цикла, но мне кажется, что это муторно, поэтому сорре, сделаю просто функцию

In [237]:
def five_models(model):
    mdl = model
    y_pred = mdl.fit(X_train, y_train).predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(accuracy)

In [238]:
rfc = RandomForestClassifier
five_models(rfc(max_depth=4, n_estimators=5))

0.930769230769


In [240]:
five_models(rfc(max_depth=21, n_estimators=40))

0.930769230769


In [241]:
five_models(rfc(max_depth=46, n_estimators=40))

0.938461538462


In [242]:
five_models(rfc(max_depth=100, n_estimators=40))

0.946153846154


In [246]:
five_models(rfc(max_depth=1, n_estimators=5))

0.938461538462


Кажется, в конфе спрашивали, может ли быть такое, что значения получаются одинаковыми. Рискну предположить, что это как раз про этот случай. 

### 3. Обучаем градиентный бустинг

3. GradientBoosting (2 балла)
    * Обучите [GradientBoostingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier)  и найдите для него наилучшие параметры с помощью кросс-валидации. (0.5 балла)<br\>
    * Изобразите таблицу важности признаков для наилучшей настройки метода. Отличается ли она от таблицы метода RandomForest. Почему? (1 балл)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода. (0.5 балла)<br\>

In [116]:
np.logspace(1, 3, 7, dtype=int)

array([  10,   21,   46,  100,  215,  464, 1000])

In [117]:
np.logspace(-1, 0, 3, dtype=float)

array([ 0.1       ,  0.31622777,  1.        ])

In [118]:
np.logspace(0, 2, 7, dtype=int)

array([  1,   2,   4,  10,  21,  46, 100])

In [120]:
gboost = GradientBoostingClassifier(random_state=RANDOM_SEED)

gboost_params = {'n_estimators': np.logspace(1, 3, 7, dtype=int),
                 'subsample': np.logspace(-1, 0, 3, dtype=float),
                 'max_depth': np.logspace(0, 2, 7, dtype=int)}

gboost_grid = GridSearchCV(gboost, gboost_params, scoring=make_scorer(accuracy_score), cv=25)

gboost = gboost_grid.fit(X_train, y_train)

gboost_grid.best_params_, gboost_grid.best_score_

/Users/anastasiarodygina/anaconda/lib/python3.6/site-packages/sklearn/ensemble/gradient_boosting.py:1093: RuntimeWarning: invalid value encountered in double_scalars
  sample_weight[~sample_mask]))
/Users/anastasiarodygina/anaconda/lib/python3.6/site-packages/sklearn/ensemble/gradient_boosting.py:1093: RuntimeWarning: invalid value encountered in double_scalars
  sample_weight[~sample_mask]))
/Users/anastasiarodygina/anaconda/lib/python3.6/site-packages/sklearn/ensemble/gradient_boosting.py:1093: RuntimeWarning: invalid value encountered in double_scalars
  sample_weight[~sample_mask]))
/Users/anastasiarodygina/anaconda/lib/python3.6/site-packages/sklearn/ensemble/gradient_boosting.py:1093: RuntimeWarning: invalid value encountered in double_scalars
  sample_weight[~sample_mask]))
/Users/anastasiarodygina/anaconda/lib/python3.6/site-packages/sklearn/ensemble/gradient_boosting.py:1093: RuntimeWarning: invalid value encountered in double_scalars
  sample_weight[~sample_mask]))
/Users/ana

({'max_depth': 2, 'n_estimators': 21, 'subsample': 1.0}, 0.88246628131021199)

Предыдущая ячейка очень долго считалась, поэтому скопирую аутпут сюда:

({'max_depth': 2, 'n_estimators': 21, 'subsample': 1.0}, 0.88246628131021199)

In [121]:
gboost_best_est = gboost.best_estimator_
gboost_best_est.get_params()

{'criterion': 'friedman_mse',
 'init': None,
 'learning_rate': 0.1,
 'loss': 'deviance',
 'max_depth': 2,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_split': 1e-07,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 21,
 'presort': 'auto',
 'random_state': 42,
 'subsample': 1.0,
 'verbose': 0,
 'warm_start': False}

In [122]:
y_pred = gboost.predict(X_test)
gboost_accuracy = accuracy_score(y_test, y_pred)

In [124]:
gboost_fimp = gboost_best_est.feature_importances_
df_fimp_3 = pd.DataFrame(gboost_fimp, columns=['важность признака'], index=df1.columns)
df_fimp_3_sorted = df_fimp_3.sort_values('важность признака', ascending=False).head(14)
df_fimp_3_sorted

важность признака
sex                      0.231480
absences                 0.208811
goout                    0.115909
G2                       0.094396
age                      0.058842
G1                       0.056551
famrel                   0.041721
reason_other             0.037560
Mjob_services            0.023388
Fjob_health              0.023132
studytime                0.022808
paid                     0.022619
Fjob_services            0.015667
guardian_other           0.015016

In [230]:
results_2 = pd.DataFrame(gboost.cv_results_)
results_2.sort_values(['mean_test_score', 'rank_test_score'], ascending=False).head()

mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
26       0.022909         0.000747         0.882466          0.884956   
11       0.068064         0.000921         0.880539          0.887686   
29       0.051553         0.000896         0.878613          0.902217   
8        0.072452         0.001230         0.876686          0.880862   
44       0.029943         0.000761         0.876686          0.902538   

   param_max_depth param_n_estimators param_subsample  \
26               2                 21               1   
11               1                100               1   
29               2                 46               1   
8                1                 46               1   
44               4                 10               1   

                                               params  rank_test_score  \
26  {'max_depth': 2, 'n_estimators': 21, 'subsampl...                1   
11  {'max_depth': 1, 'n_estimators': 100, 'subsamp...                2   
29  {'max_depth': 2, 'n_estimators': 46, 'subsampl...                3   
8   {'max_depth': 1, 'n_estimators': 46, 'subsampl...                4   
44  {'max_depth': 4, 'n_estimators': 10, 'subsampl...                4   

    split0_test_score       ...         split7_test_score  split7_train_score  \
26           0.863636       ...                  0.857143            0.887550   
11           0.863636       ...                  0.857143            0.887550   
29           0.863636       ...                  0.857143            0.905622   
8            0.818182       ...                  0.857143            0.877510   
44           0.863636       ...                  0.857143            0.903614   

    split8_test_score  split8_train_score  split9_test_score  \
26           0.904762            0.879518           0.904762   
11           0.904762            0.885542           0.904762   
29           0.904762            0.895582           0.904762   
8            0.904762            0.877510           0.857143   
44           0.904762            0.895582           0.904762   

    split9_train_score  std_fit_time  std_score_time  std_test_score  \
26            0.877510      0.002236        0.000202        0.025877   
11            0.885542      0.023286        0.000299        0.031688   
29            0.899598      0.005148        0.000292        0.025856   
8             0.879518      0.045745        0.000480        0.028057   
44            0.895582      0.002730        0.000184        0.028971   

    std_train_score  
26         0.003014  
11         0.002297  
29         0.003206  
8          0.002467  
44         0.003795  

[5 rows x 63 columns]

In [235]:
lines = [26, 11, 29, 8, 44]
best_five_params(lines=lines, df=results_2)

[array([{'max_depth': 2, 'n_estimators': 21, 'subsample': 1.0}], dtype=object),
 array([{'max_depth': 1, 'n_estimators': 100, 'subsample': 1.0}], dtype=object),
 array([{'max_depth': 2, 'n_estimators': 46, 'subsample': 1.0}], dtype=object),
 array([{'max_depth': 1, 'n_estimators': 46, 'subsample': 1.0}], dtype=object),
 array([{'max_depth': 4, 'n_estimators': 10, 'subsample': 1.0}], dtype=object)]

In [247]:
gbc = GradientBoostingClassifier
five_models(gbc(max_depth=2, n_estimators=21, subsample=1.0))

0.930769230769


In [248]:
five_models(gbc(max_depth=1, n_estimators=100, subsample=1.0))

0.930769230769


In [249]:
five_models(gbc(max_depth=2, n_estimators=46, subsample=1.0))

0.946153846154


In [250]:
five_models(gbc(max_depth=1, n_estimators=46, subsample=1.0))

0.938461538462


In [251]:
five_models(gbc(max_depth=4, n_estimators=10, subsample=1.0))

0.930769230769


### 4. Обучаем AdaBoost

4. AdaBoost (1 балл)
    * Обучите [AdaBoostClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html) и найдите для него наилучшие параметры с помощью кросс-валидации. (0.5 балла)<br\>
    * Укажите значение accuracy для пяти наилучших настроек метода. (0.5 балла)<br\>

In [125]:
adaboost = AdaBoostClassifier(random_state=RANDOM_SEED)

adaboost_params = {'n_estimators': np.logspace(1, 3, 7, dtype=int),
                   'learning_rate': np.logspace(-1, 0, 7, dtype=float),
                   'algorithm': ['SAMME', 'SAMME.R']}

adaboost_grid = GridSearchCV(adaboost, adaboost_params, scoring=make_scorer(accuracy_score), cv=25)

adaboost = adaboost_grid.fit(X_new_train, y_new_train)

adaboost_grid.best_params_, adaboost_grid.best_score_

({'algorithm': 'SAMME',
  'learning_rate': 0.10000000000000001,
  'n_estimators': 10},
 0.87090558766859349)

In [126]:
adaboost_best_est = adaboost.best_estimator_
adaboost_best_est.get_params()

{'algorithm': 'SAMME',
 'base_estimator': None,
 'learning_rate': 0.10000000000000001,
 'n_estimators': 10,
 'random_state': 42}

In [128]:
X_test.shape, y_test.shape, y_pred.shape

((130, 44), (130,), (130,))

Я облажалась и обучила адабуст на датасете с удаленными важными признакам -_-

Пока запущу с тем, что есть, потом мб поменяю

In [131]:
y_pred = adaboost.predict(X_new_test)
adaboost_accuracy = accuracy_score(y_new_test, y_pred)
adaboost_accuracy

0.93846153846153846

In [133]:
adaboost_fimp = adaboost_best_est.feature_importances_
df_fimp_4 = pd.DataFrame(adaboost_fimp, columns=['важность признака'], index=df2.columns)
df_fimp_4_sorted = df_fimp_4.sort_values('важность признака', ascending=False).head(14)
df_fimp_4_sorted

важность признака
sex                          1.0
school                       0.0
absences                     0.0
Mjob_at_home                 0.0
Mjob_health                  0.0
Mjob_other                   0.0
Mjob_services                0.0
Mjob_teacher                 0.0
Fjob_at_home                 0.0
Fjob_health                  0.0
Fjob_other                   0.0
Fjob_services                0.0
Fjob_teacher                 0.0
reason_course                0.0

Ого. Влияет только пол?!

In [134]:
adaboost = AdaBoostClassifier(random_state=RANDOM_SEED)

adaboost_params = {'n_estimators': np.logspace(1, 3, 7, dtype=int),
                   'learning_rate': np.logspace(-1, 0, 7, dtype=float),
                   'algorithm': ['SAMME', 'SAMME.R']}

adaboost_grid = GridSearchCV(adaboost, adaboost_params, scoring=make_scorer(accuracy_score), cv=25)

adaboost = adaboost_grid.fit(X_train, y_train)

adaboost_grid.best_params_, adaboost_grid.best_score_

({'algorithm': 'SAMME',
  'learning_rate': 0.10000000000000001,
  'n_estimators': 464},
 0.87861271676300579)

In [135]:
adaboost_best_est = adaboost.best_estimator_
adaboost_best_est.get_params()

{'algorithm': 'SAMME',
 'base_estimator': None,
 'learning_rate': 0.10000000000000001,
 'n_estimators': 464,
 'random_state': 42}

In [136]:
y_pred = adaboost.predict(X_test)
adaboost_accuracy = accuracy_score(y_test, y_pred)
adaboost_accuracy

0.93846153846153846

In [137]:
adaboost_fimp = adaboost_best_est.feature_importances_
df_fimp_4 = pd.DataFrame(adaboost_fimp, columns=['важность признака'], index=df1.columns)
df_fimp_4_sorted = df_fimp_4.sort_values('важность признака', ascending=False).head(14)
df_fimp_4_sorted

важность признака
sex                     0.309702
absences                0.166815
G1                      0.106437
G2                      0.099670
famrel                  0.080508
goout                   0.035995
reason_other            0.028164
age                     0.025770
Fjob_services           0.023456
Mjob_health             0.022613
freetime                0.021588
studytime               0.018425
address                 0.012982
famsize                 0.011241

In [159]:
a = [1, 2, 3, 4, 5, 3, 2]

In [160]:
a.sort()

In [161]:
a

[1, 2, 2, 3, 3, 4, 5]

In [252]:
results_3 = pd.DataFrame(adaboost.cv_results_)
results_3.sort_values(['mean_test_score', 'rank_test_score'], ascending=False).head()

mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
5        1.165530         0.048387         0.878613          0.881343   
11       0.467914         0.020686         0.878613          0.880942   
24       0.226412         0.010416         0.878613          0.881505   
18       0.492233         0.022345         0.876686          0.881664   
25       0.433914         0.020164         0.876686          0.881424   

   param_algorithm param_learning_rate param_n_estimators  \
5            SAMME                 0.1                464   
11           SAMME             0.14678                215   
24           SAMME            0.316228                100   
18           SAMME            0.215443                215   
25           SAMME            0.316228                215   

                                               params  rank_test_score  \
5   {'algorithm': 'SAMME', 'learning_rate': 0.1, '...                1   
11  {'algorithm': 'SAMME', 'learning_rate': 0.1467...                1   
24  {'algorithm': 'SAMME', 'learning_rate': 0.3162...                1   
18  {'algorithm': 'SAMME', 'learning_rate': 0.2154...                4   
25  {'algorithm': 'SAMME', 'learning_rate': 0.3162...                4   

    split0_test_score       ...         split7_test_score  split7_train_score  \
5            0.863636       ...                  0.857143            0.881526   
11           0.863636       ...                  0.857143            0.881526   
24           0.863636       ...                  0.857143            0.881526   
18           0.863636       ...                  0.857143            0.879518   
25           0.863636       ...                  0.857143            0.879518   

    split8_test_score  split8_train_score  split9_test_score  \
5            0.857143            0.879518           0.904762   
11           0.857143            0.879518           0.904762   
24           0.857143            0.881526           0.904762   
18           0.857143            0.879518           0.857143   
25           0.857143            0.875502           0.857143   

    split9_train_score  std_fit_time  std_score_time  std_test_score  \
5             0.877510      0.208980        0.011218        0.031605   
11            0.877510      0.025229        0.001708        0.031605   
24            0.877510      0.019748        0.001274        0.031605   
18            0.877510      0.069533        0.006886        0.031403   
25            0.881526      0.009297        0.001727        0.032092   

    std_train_score  
5          0.002752  
11         0.002274  
24         0.003212  
18         0.002702  
25         0.003873  

[5 rows x 63 columns]

In [253]:
lines = [26, 11, 29, 8, 44]
best_five_params(lines=lines, df=results_3)

[array([ {'algorithm': 'SAMME', 'learning_rate': 0.31622776601683794, 'n_estimators': 464}], dtype=object),
 array([ {'algorithm': 'SAMME', 'learning_rate': 0.14677992676220694, 'n_estimators': 215}], dtype=object),
 array([ {'algorithm': 'SAMME', 'learning_rate': 0.46415888336127786, 'n_estimators': 21}], dtype=object),
 array([ {'algorithm': 'SAMME', 'learning_rate': 0.14677992676220694, 'n_estimators': 21}], dtype=object),
 array([{'algorithm': 'SAMME', 'learning_rate': 1.0, 'n_estimators': 46}], dtype=object)]

In [254]:
ada = AdaBoostClassifier
five_models(ada(algorithm='SAMME', learning_rate=0.31622776601683794, n_estimators=464))

0.938461538462


In [255]:
five_models(ada(algorithm='SAMME', learning_rate=0.14677992676220694, n_estimators=215))

0.938461538462


In [256]:
five_models(ada(algorithm='SAMME', learning_rate=0.46415888336127786, n_estimators=21))

0.938461538462


In [257]:
five_models(ada(algorithm='SAMME', learning_rate=0.14677992676220694, n_estimators=21))

0.938461538462


In [258]:
five_models(ada(algorithm='SAMME', learning_rate=1.0, n_estimators=46))

0.930769230769


### 5. Выбираем лучший классификатор

5. Какой из классификаторов оказался лучше? (0.5 балла)